Add documentation for using features (#2383)
foundry-rs/starknet-foundry: docs/src/snforge-advanced-features/conditional-compilation.md

# Conditional Compilation

It is possible to build some contracts solely for testing purposes.

> This can be achieved by leveraging [Scarb features](../snforge-advanced-features/fork-testing.md).

Reviewer: It should link to scarb docs, no?

Author: Not sure what happened here, fixed now

Configuration in `Scarb.toml` is done in the same manner as described in the Scarb documentation.

> Additionally, the `snforge test` command exposes the following flags: [--features](../appendix/snforge/test.md#-f---features-features),

Reviewer: [nit] I would mention they are 1:1 Scarb's flags

[--all-features](../appendix/snforge/test.md#--all-features) and [--no-default-features](../appendix/snforge/test.md#--no-default-features).

## Contracts

> Firstly, define a contract in the `src` directory with a `#cfg(feature: (...))` attribute:

Reviewer: ```suggestion
Firstly, define a contract in the `src` directory with a `#[cfg(feature: "...")]` attribute:
```

> Firstly, define a contract in the `src` directory with a `#cfg(feature: (...))` attribute:

Reviewer: Please put an emphasis that if one wants to declare a contract in a test it has to be in the package that is being tested, not in `tests`

Reviewer: https://foundry-rs.github.io/starknet-foundry/testing/contracts-collection.html#how-contracts-are-collected

You can mention that it is like this bcs of the way snforge collects contracts and link this page

Author: Added a note about this

```rust
#[starknet::interface]
trait IMockContract<TContractState> {
    fn response(self: @TContractState) -> u32;
}

#[starknet::contract]
#[cfg(feature: 'snforge_test_only')]
mod MockContract {
    #[storage]
    struct Storage {}

    #[abi(embed_v0)]
    impl IContractImpl of super::IContract<ContractState> {
        fn response(self: @ContractState) -> u32 {
            1
        }
    }
}
```

Next, create a test with the same attribute:

```rust
#[test]
#[cfg(feature: 'snforge_test_only')]
fn test_mock_contract() {
    let (contract_address, _) = declare("MockContract")
        .unwrap()
        .contract_class()
        .deploy(@array![])
        .unwrap();
    let response_result = IContractDispatcher { contract_address }.response();
    assert(response_result == 1, '');
}

> ```

Reviewer: The cfg flag makes the example unclear imo and I think tagging tests with cfg is a bad practise for most users (I can see usecases for that, but most of them will just forget that the test didn't run). I would remove the cfg and comment in the last line of the paragraph that the test will fail if you don't pass the features flag

The `Scarb.toml` file needs to be updated:

```toml
[features]
snforge_test_only = []
```

Then, tests can be executed with:

```
snforge test --features snforge_test_only
```

If `snforge test` is run without features enabled, it won't execute the previous test or build any artifacts for the `MockContract`.

## Functions

> Similarly, we can conditionally compile some functions created in the `tests` directory:

Reviewer: It is not limited to directory here
```suggestion
Similarly, other parts of the code can also be conditionally compiled:
```

```rust
#[cfg(feature: 'snforge_test_only')]
fn mock_function() -> u32 {
    2
}

#[test]

> #[cfg(feature: 'snforge_test_only')]

Reviewer: Vide supra, bad practice imo

# Conditional Compilation

It is possible to build some contracts solely for testing purposes.
This can be achieved by leveraging [Scarb features](https://docs.swmansion.com/scarb/docs/reference/conditional-compilation.html#features).
Configuration in `Scarb.toml` is done in the same manner as described in the Scarb documentation.
Additionally, for utilizing features the `snforge test` command exposes the following flags, aligned with `scarb` flags:
[--features](../appendix/snforge/test.md#-f---features-features),
[--all-features](../appendix/snforge/test.md#--all-features) and [--no-default-features](../appendix/snforge/test.md#--no-default-features).

## Contracts

> Firstly, define a contract in the `src` directory with a `#[cfg(feature: "...")]` attribute:

Reviewer: ```suggestion
Firstly, define a contract in the `src` directory with a `#[cfg(feature: '<FEATURE_NAME>')]` attribute:
```
Mb here before xD

```rust
#[starknet::interface]
trait IMockContract<TContractState> {
    fn response(self: @TContractState) -> u32;
}

#[starknet::contract]

> #[cfg(feature: 'snforge_test_only')]

Reviewer 2: A comment here explaining that the name can be `anything` could be useful. Or use a different feature name, that doesn't look like a build in thing in forge.

Author: Renamed to `enable_for_tests`

mod MockContract {
    #[storage]
    struct Storage {}

    #[abi(embed_v0)]
    impl IContractImpl of super::IContract<ContractState> {
        fn response(self: @ContractState) -> u32 {
            1
        }
    }
}
```

> 📝 **Note**

> > To declare mock contracts in tests, these contracts should be defined within the package and not in the `tests` directory.

Reviewer 2: ```suggestion
> To use mock contracts in tests, these contracts must be defined within the package and not in the `tests` directory.
```

Reviewer: It is specifically about declaring. You can use contract module defined in `tests` directory in tests

> This requirement is due to the way snforge [collects contracts](../testing/contracts-collection.md).

Next, create a test that uses the above contract:

```rust
#[test]
fn test_mock_contract() {
    let (contract_address, _) = declare("MockContract")
        .unwrap()
        .contract_class()
        .deploy(@array![])
        .unwrap();
    let response_result = IContractDispatcher { contract_address }.response();
    assert(response_result == 1, '');
}
```

> The `Scarb.toml` file needs to be updated:

Reviewer 2: ```suggestion
After that, update the `Scarb.toml` file so it includes the following lines:
```

Reviewer: The form is inconsistent if you do it

Reviewer 2: `The `Scarb.toml` file needs to be updated so it includes the following lines:`?

```toml
[features]
snforge_test_only = []
```

> Then, tests can be executed with:

Reviewer 2: ```suggestion
Then, to use the contract in tests `snforge test` must be provided with a flag defined above:
```

```
snforge test --features snforge_test_only
```

> If `snforge test` is run without features enabled, it won't build any artifacts for the `MockContract` and all tests that use this contract will fail.

Reviewer: ```suggestion
If `snforge test` is run without the above feature enabled, it won't build any artifacts for the `MockContract` and all tests that use this contract will fail.
```

> If `snforge test` is run without the above feature enabled, it won't build any artifacts for the `MockContract` and all tests that use this contract will fail.

Reviewer 2: This could be a warning or a note maybe?

Author: Changed to a note

## Functions

> Similarly, other parts of the code can also be conditionally compiled:

Reviewer 2: ```suggestion
Features are not limited to conditionally compiling contracts and can be used with other parts of the code, like functions:
```
